pkg/modprovider/views.go

```diff
@@ -0,0 +1,242 @@
+
+	// resource is replaced by deleting the old copy then creating a new copy
+	case tfsandbox.ReplaceDestroyBeforeCreate:
+		return pulumirpc.ViewStep_CREATE_REPLACEMENT // TODO is this right?
+
+	case tfsandbox.Create:
+		return pulumirpc.ViewStep_CREATE
+
+	// TODO is this always right? Currently only supporting refresh-to-Read.
+	case tfsandbox.Read:
+		return pulumirpc.ViewStep_REFRESH
+
+	case tfsandbox.Delete:
+		return pulumirpc.ViewStep_DELETE
+	case tfsandbox.Forget:
+		contract.Failf("Forget operations are not yet supported")
+	}
+	contract.Failf("Unrecognized changeKind: %v", changeKind)
+	var zero pulumirpc.ViewStep_Op
+	return zero
+}
+
+// Starting with very basic error checks for starters. It should be possible to extract more information from TF.
+func viewStepStatusCheck(
+	changeKind tfsandbox.ChangeKind,
+	finalState *tfsandbox.ResourceState, // may be nil when planning or failed to create
+) error {
+
+	switch changeKind {
+
+	// All these operations when successful imply the resource must exist in the final state.
+	case tfsandbox.NoOp, tfsandbox.Update, tfsandbox.Create,
+		tfsandbox.Replace, tfsandbox.ReplaceDestroyBeforeCreate:
+		if finalState == nil {
+			return errors.New("resource operation failed")
+		}
+
+	// These operations if successful imply the resource must not exist in the final state.
+	case tfsandbox.Delete, tfsandbox.Forget:
+		if finalState != nil {
+			return errors.New("resource operation failed")
+		}
+	}
+
+	return nil
+}
+
+func viewStepForResource(
+	packageName packageName,
+	rplan *tfsandbox.ResourcePlan,
+	priorState *tfsandbox.ResourceState, // may be nil in operations such as create
+	finalState *tfsandbox.ResourceState, // may be nil when planning or failed to create
+) *pulumirpc.ViewStep {
+
+	ty := childResourceTypeToken(packageName, rplan.Type()).String()
+	name := childResourceName(rplan)
+
+	var oldViewState, newViewState *pulumirpc.ViewStepState
+	if finalState != nil {
+		newViewState = viewStepState(packageName, finalState)
+	} else {
+		newViewState = viewStepState(packageName, rplan)
+	}
+
+	if priorState != nil {
+		oldViewState = viewStepState(packageName, priorState)
+	}
+
+	step := &pulumirpc.ViewStep{
+		Status: pulumirpc.ViewStep_OK,
+		Name:   name,
+		Type:   ty,
+
+		Op:  viewStepOp(rplan.ChangeKind()),
+		Old: oldViewState,
+		New: newViewState,
+
+		// TODO translate TF diff details to Pulumi view diff details.
+		//
+		// Keys:            []string{},                           // need to attribute replacement plans to properties here
+		// Diffs:           []string{},                           // need to provide an approximation of DetailedDiff here
+		// DetailedDiff:    map[string]*pulumirpc.PropertyDiff{}, // need to populate this
+		// HasDetailedDiff: true,
+	}
+
+	if err := viewStepStatusCheck(rplan.ChangeKind(), finalState); err != nil {
+		// TODO is this right? It is not so much that the resource failed partially, it is that the entire
```

Comment:
@justinvp wonder if you could explain this bit, if a "view" totally failed, do we mark it as PARTIAL_FAILURE?

Reply:
I'm not very family with partial failures, but they seem pretty rare from spelunking usage. I'm guessing we don't even need to make use of them for modules/views? Does the bridge even ever return partial failures for bridged providers?

Generally, if a Module has failed, it should error in the usual way from its CRUD operation, and any child views that have failed would have their `step.Error` set to the error message (and leave status blank for the default value, which is StatusOK).

Reply:
It seems nice to attribute errors to a particular view when you have 30 of those.